### PR TITLE
New Apache Camel Kafka Component

### PR DESCRIPTION
**Pull request by khajurkar**
New component - Camel Kafka Producer, which uses Kafka Asynchronous. The existing Camel-Kafka 2.15 component, uses single low level Producer The kafka.javaapi.Producer holds a global lock, before sending every single message

The new 0.8.2.x Async Producer by Kafka Java API, let ones tune the locking and latency by defining the batchSize and lingering time as configurables. linger.ms = 0, means as good as flushing the message buffer as soon as it is received.

### CODE REVIEW

File: fk-camel-kafka-async-component.iml [hunk 1]
```diff
@@ -0,0 +1,52 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (vivekagarwa) [inline fk-camel-kafka-async-component.iml]: Remove this files from repo

File: kafka.iml [hunk 2]
```diff
@@ -0,0 +1,46 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (vivekagarwa) [inline kafka.iml]: Remove this files from repo

File: release.sh [hunk 3]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
+
+set -x
+cat pom.xml | grep -F '<version>' | head -1 | sed -e 's/<[^/]*>//g' -e 's/<.*>//g' -e 's/ //g' | grep -q '\-t9$'
```
Comment (vivekagarwa) [inline release.sh:4]: This is not required

File: src/main/java/flipkart/dsp/camel/AsyncKafkaComponent.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
+/**
```
Comment (vivekagarwa) [inline src/main/java/flipkart/dsp/camel/AsyncKafkaComponent.java]: Remove license

File: src/main/java/flipkart/dsp/camel/AsyncKafkaProducerConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+public class AsyncKafkaProducerConfiguration {
+
+    @UriParam(defaultValue = "100")
+    private Integer lingerMs = 100;
+
+    @UriParam
+    private String clientId;
+
+    @UriParam(defaultValue = "all")
+    private String acks;
```
Comment (vivekagarwa) [inline src/main/java/flipkart/dsp/camel/AsyncKafkaProducerConfiguration.java]: Although Kafka producer will consider default value for acks as all. But just to be on safer side as we are declaring all to be defaultValue we should set it default as all on our end also.

File: src/main/java/flipkart/dsp/camel/AsyncKafkaProducerConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+    private String topic;
+
+    @UriParam
+    private String keySerializer;
+
+    @UriParam
+    private String valueSerializer;
+
+    @UriParam(defaultValue = "5")
+    private Integer retries;
```
Comment (vivekagarwa) [inline src/main/java/flipkart/dsp/camel/AsyncKafkaProducerConfiguration.java]: Same as acks.

File: src/main/java/flipkart/dsp/camel/AsyncKafkaProducerConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+    @UriParam(defaultValue = "1000")
+    private Integer reconnectBackoffMs;
```
Comment (vivekagarwa) [inline src/main/java/flipkart/dsp/camel/AsyncKafkaProducerConfiguration.java]: same as acks.

File: src/main/java/flipkart/dsp/camel/AsyncKafkaProducerConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+    @UriParam(defaultValue = "1000")
+    private Integer retryBackoffMs;
```
Comment (vivekagarwa) [inline src/main/java/flipkart/dsp/camel/AsyncKafkaProducerConfiguration.java]: same as acks.

File: release.sh (hunk 3 above)
Comment (khajurkar) [inline release.sh:4]: Didn't get this quite. I just blindly used it from the existing repos

File: fk-camel-kafka-async-component.iml (hunk 1 above)
Comment (khajurkar) [inline fk-camel-kafka-async-component.iml]: done

File: kafka.iml (hunk 2 above)
Comment (khajurkar) [inline kafka.iml]: done

File: src/main/java/flipkart/dsp/camel/AsyncKafkaProducerConfiguration.java (hunk 5 above)
Comment (khajurkar) [inline src/main/java/flipkart/dsp/camel/AsyncKafkaProducerConfiguration.java]: done. fixed

File: src/main/java/flipkart/dsp/camel/AsyncKafkaProducerEndpoint.java [hunk 9]
```diff
@@ -0,0 +1,168 @@
+/**
```
Comment (vivekagarwa) [inline src/main/java/flipkart/dsp/camel/AsyncKafkaProducerEndpoint.java:1]: Remove license

File: src/test/java/flipkart/dsp/camel/AsyncKafkaProducerEndpointTest.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+import java.net.URISyntaxException;
+
+import static flipkart.dsp.camel.ComponentTestConstants.*;
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Created by bhushan.sk on 31/08/15.
```
Comment (vivekagarwa) [inline src/test/java/flipkart/dsp/camel/AsyncKafkaProducerEndpointTest.java:13]: Remove comment

File: src/main/java/flipkart/dsp/camel/KafkaConstants.java [hunk 11]
```diff
@@ -0,0 +1,10 @@
+package flipkart.dsp.camel;
+
+/**
+ * Created by bhushan.sk on 25/08/15.
```
Comment (vivekagarwa) [inline src/main/java/flipkart/dsp/camel/KafkaConstants.java:4]: Remove comment

File: src/main/java/flipkart/dsp/camel/AsyncKafkaProducerConfiguration.java [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+
+    public AsyncKafkaProducerConfiguration() {
+//        setClientId((this.getTopic() != null ?
```
Comment (vivekagarwa) [inline src/main/java/flipkart/dsp/camel/AsyncKafkaProducerConfiguration.java:51]: Remove comment

File: src/main/java/flipkart/dsp/camel/AsyncKafkaProducerConfiguration.java [hunk 13]
```diff
@@ -0,0 +1,166 @@
+package flipkart.dsp.camel;
+
+/**
```
Comment (vivekagarwa) [inline src/main/java/flipkart/dsp/camel/AsyncKafkaProducerConfiguration.java:3]: Remove comment

File: src/main/java/flipkart/dsp/camel/AsyncKafkaProducer.java [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+import org.apache.camel.impl.DefaultProducer;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+
+import java.util.Properties;
+import java.util.concurrent.Future;
+
+/**
```
Comment (vivekagarwa) [inline src/main/java/flipkart/dsp/camel/AsyncKafkaProducer.java:18]: Remove comment

File: src/test/java/flipkart/dsp/camel/AsyncKafkaProducerTest.java [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+
+
+    void sendMessageViaKafkaLowlevelKafkaProducer() throws Exception {
+
+        this.kafkaProducerEndpoint.setTopic(TOPIC);
+        in.setBody("My name is foo");
+        when(exchange.getIn()).thenReturn(in);
+        this.kafkaProducer.process(exchange);
+
+        verify(kafkaProducer.producer, times(1)).send(Matchers.any(ProducerRecord.class));
```
Comment (vivekagarwa) [inline src/test/java/flipkart/dsp/camel/AsyncKafkaProducerTest.java:81]: seems a little wrong. The function name suggests that this function will just send, while it is verifying also.

**Comment by vivekagarwa** [comment]
Minor comments

**Pull request merged**